R/add_coordinates.R

```diff
@@ -1,63 +1,88 @@
+#' @author Pawel Piatkowski, Zhian Kamvar
 #' 
 #' @examples
-#' flows <- do.call(make_epiflows, Mex_travel_2009)
-#' flows <- add_coordinates(flows)
-#' flows$locationsdata
+#' 
+#' # Setting up the data
+#' data(Mex_travel_2009)
+#' flows   <- Mex_travel_2009[[1]]
+#' outflow <- setNames(flows[["MEX"]], rownames(flows))
+#' inflow  <- unlist(flows["MEX", , drop = TRUE])
+#' ef <- epiflows(inflow, outflow, focus = "MEX", locations = Mex_travel_2009[[2]])
+#' 
+#' # generating random coordinates for demonstration
+#' dummy_coords <- data.frame(lon = runif(41, min = -95, max = 95),
+#'                            lat = runif(41, min = -95, max = 95))
+#' print(ef)
+#' print(flows <- add_coordinates(ef, coords = dummy_coords))
+#' get_coords(flows)
+#' 
+#' \dontrun{
+#'   # You can use google maps' geocode functionality if you have a decent 
+#'   # internet connection
+#'   flows <- add_coordinates(ef, loc_column = "country")
+#' }
 #' 
 #' @export
-add_coordinates <- function(ef,
-                            loc_column = "country",
-                            lon_lat_columns = c("lon", "lat"),
+add_coordinates <- function(ef, coords = c("lon", "lat"), loc_column = "id",
                             overwrite = FALSE) {
-  if (!"epiflows" %in% class(ef)) {
-    stop("`ef` must be an object of class epiflows")
+  if (!inherits(ef, "epiflows")) {
+    efprint <- as.character(deparse(substitute(ef)))
+    stop(sprintf("%s must be an object of class epiflows", efprint))
   }
-  if (!loc_column %in% names(ef$locationsdata)) {
-    stop(sprintf("`%s` is not a valid column name", loc_column))
+  if (!is.null(ef$vars$coords) && !overwrite) {
+    stop("coordinates are present in the object. Use `overwrite = TRUE` to replace them.")
   }
-  if (!is.data.frame(lon_lat_columns) && !is.character(lon_lat_columns) || length(lon_lat_columns) != 2 ) {
-    stop("`lon_lat_columns` should contain exactly two character strings")
-  }
-  
-  if(is.data.frame(lon_lat_columns)){
-    if(ncol(lon_lat_columns) != 2){
-      stop("The data frame `lon_lat_columns` should contain exactly two columns specifying the longitude and latitude coordinates")
+  if (is.data.frame(coords)) {
+    if (ncol(coords) != 2) {
+      stop("The data frame `coords` should contain exactly two columns specifying the longitude and latitude coordinates")
+    }
+    if (all(c("lon", "lat") %in% tolower(names(coords)))) {
+      # Ensuring they are in the correct order if they are named lon/lat
+      # otherwise, we just assume they are.
+      names(coords) <- tolower(names(coords))
+      coords <- coords[c("lon", "lat")]
+    }
+    if (overwrite && !is.null(ef$vars$coords)) {
+      names(coords)               <- ef$vars$coords
+      ef$linelist[ef$vars$coords] <- coords
+    } else {
+      ef$linelist    <- cbind(get_locations(ef), coords) 
+      ef$vars$coords <- setNames(names(coords), c("lon", "lat"))
+    }
+  } else {
+    if (!is.character(coords) || length(coords) != 2) {
```

Comment:
I increasingly assume I don't know how operator priority works, as I once had a nightmare juggling between R and C++; so I'd write   
```
    if (!is.character(coords) || (length(coords) != 2L)) {
```
but your call

Reply:
This scheme will work. I guarantee it. We can talk about your nightmares later.